spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+      Monitoring monitoring,
+      String customTypeNamespace,
+      String projectName, String applicationName, String applicationVersion,
+      MeasurementFilter measurementFilter) {
+    if (projectName.isEmpty()) {
+      throw new IllegalStateException(
+          "You need to configure a stackdriver.projectName.");
+    }
+    this.service = monitoring;
+    this.customTypePrefix = String.format("custom.googleapis.com/%s/",
+                                          customTypeNamespace);
+    this.projectResourceName = "projects/" + projectName;
+    this.applicationName = applicationName;
+    this.measurementFilter = measurementFilter;
+
+    UUID uuid = UUID.randomUUID();
+    byte[] uuidBytes = new byte[16];
+    addLong(uuidBytes, 0, uuid.getLeastSignificantBits());
+    addLong(uuidBytes, 8, uuid.getMostSignificantBits());
+    setInstanceId(java.util.Base64.getEncoder().encodeToString(uuidBytes));
+  }
+
+  /**
+   * Helper function to encode a value into buffer when constructing a UUID.
+   */
+  private void addLong(byte[] buffer, int offset, long value) {
+    for (int i = 0; i < 8; ++i) {
+      buffer[i + offset] = (byte)(value >> (8 * i) & 0xff);
+    }
+  }
+
+  /**
+   * [Re]bind the instanceId used to identify the TimeSeries data coming from
+   * this particular instance.
+   */
+  public void setInstanceId(String id) {
+    log.info("Configuring {}={}", INSTANCE_LABEL, id);
+    instanceId = id;
+  }
+    
+  /**
+   * Only useful for testing.
+   */
+  public String getInstanceId() {
+    return instanceId;
+  }
+
+  /**
+   * Lookup all the pre-existing custom metrics for spinnaker so that we
+   * know if we need to create new metric descriptors or not.
+   */
+  public void initKnownDescriptors()
+      throws HttpResponseException, IOException {
+    ListMetricDescriptorsResponse response =
+      service.projects().metricDescriptors().list(this.projectResourceName)
+      .execute();
+
+    for (MetricDescriptor descriptor : response.getMetricDescriptors()) {
+      if (descriptor.getType().startsWith(customTypePrefix)) {
+        knownDescriptors.put(descriptor.getType(), descriptor);
+      }
+    }
+  }
+
+  /**
+   * Inject a hints about one or more types.
+   *
+   * Hints are only used if and when we need to create a type descriptor for
+   * a type. This happens at most once over the lifetime of a system (not just
+   * a process invocation). Adding hints is a pre-emtive workaround for known
+   * inconsistent uses for given types where an incomplete descriptor may
+   * otherwise be inferred.
+   *
+   * Note that the name of the hint is just the Spectator name, not the
+   * Stackdriver type name. That is, it does not yet contain the customTypePrefix.
+   * This is for [assumed] convienence specifying the hints.
+   */
+  public void addCustomDescriptorHints(List<CustomDescriptorHint> hints) {
+    for (CustomDescriptorHint hint : hints) {
+      addLabelHints(hint.name, hint.labels);
```

Comment:
If `addLabelHints` is simpler and this just calls it, why do we need both methods?